### PR TITLE
add forward energy edge filter for seam carving

### PR DESCRIPTION
**Pull request by axu2**
## Description
Looking for code review as my first real open source contribution!

For more details see the original issue #3082

I added an example in the pr.

## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](https://scikit-image.org/docs/dev/contribute.html)]

## References
Implements this paper:
http://www.faculty.idc.ac.il/arik/SCWeb/vidret/index.html

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by axu2** [comment]
Also, unsure why CI failed?

**Comment by hmaarrfk** [comment]
you broke all of python!!!!

**Comment by hmaarrfk** [comment]
probably some bug in a pytest 4.1, you can try to exclude it in the test depencendies

see the `requirements` directory

**Comment by axu2** [comment]
Looks ok now.

**Comment by jni** [comment]
@axu2 nice work! The code is very clear.

Could I ask you to improve the example to discuss *why* you would use forward energy over Sobel? My idea for what the example should look like is:

- old example (with the rocket)
- start of a second section: what about this other image? (bench)
- this doesn't look great, and the reason is [why sobel fails in this use case]
- to help with this, Rubinstein et al proposed the *forward energy filter*, which measures [short description of the forward energy filter]
- this results in this better output [side by side bench results with Sobel and forward energy].

The short description should also go in the docstring of the forward energy filter, by the way. It would help me in my review to know *what* the filter is trying to measure. Currently I can read the steps but I'm missing the big picture.

Finally, we use the [NumPy docstring format](https://docs.scipy.org/doc/numpy/docs/howto_document.html) rather than the ReStructuredText format.

Thanks again, I'll give this another round once you've added those descriptions!

**Comment by axu2** [comment]
@jni How's that?

File: doc/examples/transform/plot_seam_carving.py [hunk 1]
```diff
@@ -83,3 +139,63 @@
...
+######################################################################
+
+out = seam_carve(img, energy, 'vertical', 400)
+plt.figure()
+plt.title('Resized using Seam Carving')
+plt.imshow(out)
+
+######################################################################
+# Seam carving using the Sobel edge filter distorts the person
+# and unncessarily preserves pixels along the bench and the fence
```
Comment (jni) [inline doc/examples/transform/plot_seam_carving.py]:
```suggestion
# and unnecessarily preserves pixels along the bench and the fence
```

File: skimage/filters/edges.py [hunk 2]
```diff
@@ -534,3 +534,73 @@ def laplace(image, ksize=3, mask=None):
...
+        mL = np.roll(mU, 1)
+        mR = np.roll(mU, -1)
+
+        mULR = np.array([mU, mL, mR])
+        cULR = np.array([cU[i], cL[i], cR[i]])
+        mULR += cULR
+
+        argmins = np.argmin(mULR, axis=0)
+        m[i] = np.choose(argmins, mULR)
+        energy[i] = np.choose(argmins, cULR)
```
Comment (jni) [inline skimage/filters/edges.py]: I'm still unclear on what the implementation is doing. Probably some comments to map these arrays to the intuition behind the method (the forward energy) would be useful.

File: skimage/filters/edges.py [hunk 3]
```diff
@@ -534,3 +534,73 @@ def laplace(image, ksize=3, mask=None):
...
     result = convolve(image, laplace_op)
     return _mask_filter_result(result, mask)
+
+
+def forward_energy(image, mode, mask=None):
+    """Find the edge magnitude using forward energy for seam carving.
+
+    Depending on the direction in `mode`, determines the magnitude of
+    each pixel based on the new edges created after removing a seam
+    containing that pixel.
```
[the remark this answers is not in the document]
Comment (axu2) [reply skimage/filters/edges.py]: https://www.youtube.com/watch?v=AJtE8afwJEg

This video would probably be helpful for understanding.

**Comment by stefanv** [comment]
I'm sorry, folks, but sadly @jni is right.  This algorithm [is patented](https://patents.google.com/patent/US8213745B2/en) and will need to go.

I'm OK with someone taking the code and making it available somewhere else.  It would be good to make explicit mention of the issue (although our license does not require that, specifically).

**Comment by jni** [comment]
@axu2 I'm sorry, but for the reasons stated above, we cannot accept this code into scikit-image (and the original code is in the process of being removed). It's a shame because it is good work. If you host it somewhere public, feel free to post a link to this thread for future reference.

**Comment by grlee77** [comment]
As detailed in https://github.com/scikit-image/scikit-image/issues/3082#issuecomment-1083631678, I think this patent is now expired for non-payment of maintenance fees, so we can possibly resurrect this!

**Pull request closed without merging**